Dockerfile

```diff
@@ -0,0 +1,22 @@
+FROM node:current-alpine AS base
+WORKDIR /base
+COPY package*.json /.
+RUN npm install
+COPY . .
+
+FROM base AS build
+ENV NODE_ENV=production
+WORKDIR /build
+COPY --from=base /base ./
+RUN npm run build
+
+FROM node:current-alpine AS production
+ENV NODE_ENV=production
+WORKDIR /app
+COPY --from=build /build/package*.json ./
+COPY --from=build /build/.next ./.next
+COPY --from=build /build/public ./public
+RUN npm install next
+
+EXPOSE 3000
+CMD npm run start
```

Comment:
So, ran it locally and it all works, but I'm wondering if you can run me through the steps here. 
When I've set up Dockerfiles previously they've been less involved than this.

For example: https://github.com/pcraig3/vacs/blob/main/Dockerfile

<details>
<summary>click to open</summary>

```
# use node alpine
FROM node:12-alpine

# set env var "PORT" to 3000
ENV PORT 3000

# Create app directory
RUN mkdir -p /app
WORKDIR /app

# Copy source files
COPY package*.json /app/
COPY . /app/

# Install  dependencies
RUN npm install --production --silent

# Building app
RUN npm run build

EXPOSE 3000
CMD ["npm", "run", "start:production"]
```
</details>

I guess my question is why do we say `FROM` 3 times?

In the repo I linked to, I also have a `.dockerignore` file so that we don't copy everything.

Reply:
Just to be clear, I'm not requesting changes, just wondering what's up.

Reply:
Sure! This is kind of new to me as well but from my understanding those multiple FROM statements are to make a multi-stage build that reduces the final image size by only copying the most necessary files through the stages and leaving what's not needed behind. So in this case the node_modules are left out of the image. Its all based on this example here:

https://dev.to/chrsgrrtt/dockerising-a-next-js-project-1ck5

Reply:
This is actually very smart. If nothing new is installed the npm install step will be skipped and go straight to building. I like this well done